drivers/net/ethernet/intel/idpf/idpf_eth_idc.h

```diff
@@ -0,0 +1,199 @@
+ * @max_sg_bufs_per_tx_pkt: Max number of scatter gather buffers that can be
+ *			    sent per transmit packet without needing to be
+ *			    linearized.
+ * @min_sso_packet_len: Min packet length supported by device for single
+ *	segment offload
+ * @q_info: Max queue information
+ * @crc_enable: Enable CRC insertion offload
+ */
+struct idpf_eth_idc_auxiliary_dev_caps {
+	__le32 csum_caps;
+	__le32 seg_caps;
+	__le32 hsplit_caps;
+	__le32 rsc_caps;
+	__le64 rss_caps;
+	__le64 other_caps;
+	__le16 max_tx_hdr_size;
+	u8 max_sg_bufs_per_tx_pkt;
+	u8 min_sso_packet_len;
+	struct idpf_max_q q_info;
+	bool crc_enable;
+};
+
+/**
+ * struct idpf_eth_idc_event - ethernet idc events
+ * @event_data: Event data
+ * @event_code: Event code
+ */
+struct idpf_eth_idc_event {
+	void *event_data;
+	enum idpf_eth_idc_event_code event_code;
+};
+
+struct idpf_eth_idc_dev_info;
+/**
+ * struct idpf_eth_idc_ops - Ethernet driver ops to main driver
+ * @event_send: Sends ethernet event to main driver
+ * @virtchnl_send: Sends Virtchnl messages call to main driver
+ * @intr_reg_init: Initialize interrupt registers call to main driver
+ * @intr_init_vec_idx: Initialize the vector indexes call to main driver
+ * @req_rel_vec_idx: Release vector indexes call to main driver
+ */
+struct idpf_eth_idc_ops {
+	void (*event_send)(struct idpf_eth_idc_dev_info *dev_info,
+			   struct idpf_eth_idc_event *event);
+	size_t (*virtchnl_send)(struct idpf_eth_idc_dev_info *dev_info,
+				struct idpf_vc_xn_params *params);
+	int (*intr_reg_init)(struct idpf_eth_idc_dev_info *dev_info,
+			     u16 num_vecs, struct idpf_q_vector *q_vectors,
+			      u16 *q_vector_idxs);
+	int (*intr_init_vec_idx)(struct idpf_eth_idc_dev_info *dev_info,
+				 u16 num_vecs,
+				 struct idpf_q_vector *q_vectors,
+				 u16 *q_vector_idxs);
+	int (*req_rel_vec_idx)(struct idpf_eth_idc_dev_info *dev_info,
+			       u16 *num_vectors,
+			       struct idpf_vector_info *vec_info,
+			       struct msix_entry *msix_table);
+};
+
+struct idpf_eth_idc_auxiliary_driver;
+
+/**
+ * struct idpf_eth_shared - Common Device data struct shared with eth
+ * @pdev: PCI device struct given on probe
+ * @msg_enable: Debug message level enabled
+ * @hw_addr: Hardware address for use by Ethernet driver
+ */
+struct idpf_eth_shared {
+	struct pci_dev *pdev;
+	struct idpf_eth_idc_ops eth_idc_ops;
+	u32 msg_enable;
+	void __iomem *hw_addr;
+};
+
+/**
+ * struct idpf_eth_idc_dev_info - Ethernet driver's device information struct
+ * @eth_shared: Ethernet shared data struct
+ * @vport_type: Vport type
+ * @caps: Auxiliary device capabilities
+ * @start_probe_post_init: Starts probe post initialization 
+ * @vport_id: Vport ID of device
+ * @idx: Index number of auxiliary device
+ */
+struct idpf_eth_idc_dev_info {
+	struct idpf_eth_shared *eth_shared;
+	enum idpf_vport_type vport_type;
+	struct idpf_eth_idc_auxiliary_dev_caps caps;
+	bool start_probe_post_init;
```

Comment:
??aux probe is called by aux bus right??this is not clear...

Reply:
aux probe is the side effect of calling auxiliary_device_add. The device add cannot be performed again during reset. Hence the need for explicit invocation.

Reply:
There is also a workaround in the driver to run the pre-init portion of the code sequentially which can be done only by controlling how we initialize the aux driver. When card's firmware supports parallel execution, we can remove this dependency.

Reply:
Moved to the eth adapter.